### PR TITLE
Allow Actix to be started on an external CurrentThread runtime

### PR DESCRIPTION
**Pull request by GeorgeHahn**
This PR includes the minimum set of changes to enable running Actix on an external runtime.

Example usage:

```
let rt = tokio_current_thread::CurrentThread.handle();
let system = actix::System::new_async("async system", rt.to_owned());
rt.spawn(system.run_nonblocking().map_err(|e| error!("{}", e)))
  .map_err(|e| error!("Could not start actix async: {}", e))
```

### CODE REVIEW

**Comment by fafhrd91** [comment]
i think `AsyncSystemRunner` should be internal impl. System::new_async should just return impl Future

File: actix-rt/src/builder.rs [hunk 1]
```diff
@@ -127,6 +151,48 @@ impl Builder {
...
+}
+
+impl AsyncSystemRunner {
+    /// This function will start event loop and returns a future that
+    /// resolves once the `System::stop()` function is called.
+    pub fn run_nonblocking(self) -> Box<Future<Item = (), Error = io::Error> + Send + 'static> {
+        let AsyncSystemRunner { stop, .. } = self;
+
+        // run loop
+        Box::new(future::ok(())
```
Comment (fafhrd91) [inline actix-rt/src/builder.rs]: i think you can collapse most of .and_then and put everything into `future::lazy()`

File: actix-rt/src/system.rs [hunk 2]
```diff
@@ -55,6 +56,14 @@ impl System {
         Self::builder().name(name).build()
     }
 
+    #[allow(clippy::new_ret_no_self)]
+    /// Create new system using provided CurrentThread Handle.
+    ///
+    /// This method panics if it can not spawn system arbiter
+    pub fn new_async<T: Into<String>>(name: T, executor: Handle) -> AsyncSystemRunner {
```
Comment (fafhrd91) [inline actix-rt/src/system.rs]: i think this should be something like `run_in_executor`

**Comment by GeorgeHahn** [comment]
Thanks for the review! I've made the changes you requested & tested them out locally.

**Comment by fafhrd91** [comment]
thanks!

**Comment by GeorgeHahn** [comment]
@mmacedoeu Nothing in-tree, but here's an updated version of the example at the top of this PR:

```
let rt = tokio_current_thread::CurrentThread.handle();
let system = actix::System::run_in_executor("async system");
rt.spawn(system.map_err(|e| error!("Error from actix system: {}", e)));
```

Actix will be fully functional after this code runs.

**Comment by mmacedoeu** [comment]
I think when you pass the handle to run_in_executor it already spawn system.
I am trying to send a message using send from an Addr returned after a SyncArbiter::start(...) then spawn send Request as usual but Message is not processed. Could you elaborate an example where a message is spawned to an Addr. I think something is missing here.

**Comment by GeorgeHahn** [comment]
Passing the handle into `run_in_executor` allows actix to spawn the system arbiter but does not run the actix system. The returned future must be spawned on a tokio runtime.

Here's a full example:
`Cargo.toml`
```toml
[package]
name = "actix-test"
version = "0.1.0"
authors = ["George Hahn <george.hahn.vhs@gmail.com>"]
edition = "2018"

[dependencies]
actix = "0.8"
tokio = "0.1"
tokio-current-thread = "0.1"
log = "*"
futures = "0.1"
```

`main.rs`
```rust
use actix::prelude::*;
use futures::Future;
use log::error;
use tokio_current_thread::CurrentThread;

struct Ping(usize);

impl Message for Ping {
    type Result = usize;
}

struct MyActor {
    count: usize,
}

impl Actor for MyActor {
    type Context = Context<Self>;
}

impl Handler<Ping> for MyActor {
    type Result = usize;

    fn handle(&mut self, msg: Ping, _: &mut Context<Self>) -> Self::Result {
        self.count += msg.0;
        self.count
    }
}

fn main() {
    let mut rt = CurrentThread::new();
    let handle = rt.handle();
    let system = actix::System::run_in_executor("async system", rt.handle());
    handle
        .spawn(system.map_err(|e| error!("Error from actix system: {}", e)))
        .expect("start actix");

    let addr = MyActor { count: 10 }.start();
    let res = addr.send(Ping(10));

    handle
        .spawn(
            res.map(|res| {
                println!("RESULT: {}", res == 20);
            })
            .map_err(|_| ()),
        )
        .unwrap();

    rt.run().unwrap();
}
```

**Pull request merged**